### PR TITLE
02 docker and postresql

### PR DESCRIPTION
**Pull request by aziarnik**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+FROM node:lts-alpine
+WORKDIR /usr/src
+COPY "package.json" "./"
+COPY "package-lock.json" "./"
+RUN npm install --omit=dev
+RUN apk update && apk add --no-cache git
+COPY ./config /usr/src/config
+COPY --from=builder /usr/src/dist /usr/src/dist
+CMD ["npm", "run", "start-docker-prod"]
```
Comment (Miszczu2137) [inline Dockerfile]: Why this code is duplicated ?

File: config/default.json [hunk 2]
```diff
@@ -1,3 +1,10 @@
 {
-    "port": 4000
+    "port": 4000,
+    "db": {
```
Comment (Miszczu2137) [inline config/default.json]: This configuration is build for one DB, I'd prepare configuration for more than one DB.

File: config/default.json [hunk 3]
```diff
@@ -1,3 +1,10 @@
...
+        "user": "admin",
+        "host": "localhost",
```
Comment (Miszczu2137) [inline config/default.json]: Rather than user, host, password I'd pass already build connection string.

File: docker-compose.debug.yml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+version: '3.4'
+
+services:
```
Comment (Miszczu2137) [inline docker-compose.debug.yml]: docker-compose.debug.yml ? Why it's called debug ?

File: package.json [hunk 5]
```diff
@@ -4,12 +4,17 @@
...
+    "start-dev": "export NODE_ENV=default && npm run build && nodemon dist/app.js",
+    "start": "export NODE_ENV=production && npm run build && node dist/app.js",
+    "start-docker-dev": "export NODE_ENV=local-docker && npm run build && nodemon --inspect=0.0.0.0:9229 dist/app.js",
+    "start-docker-prod": "export NODE_ENV=docker && node dist/app.js",
+    "build-docker-prod": "export NODE_ENV=docker && npm run build",
     "lint": "eslint . --ext .ts",
-    "prettier-format": "prettier --config .prettierrc 'src/**/*.ts' --write"
+    "prettier-format": "prettier --config .prettierrc 'src/**/*.ts' --write",
+    "docker-compose-up": "docker builder prune && docker compose --env-file .env.prod -f 'docker-compose.yml' up -d --build",
+    "docker-compose-dev-up": "docker compose --env-file .env.dev -f 'docker-compose.debug.yml' up -d --build"
```
Comment (Miszczu2137) [inline package.json]: Please add readme file with information how to setup and run this application.

File: src/app.ts [hunk 6]
```diff
@@ -1,14 +1,32 @@
...
+app.use(versionRoute);
+app.use(fakeRoute);
+
+app.use(errorHandler);
 
-app.listen(port, () => {
-  console.log(`Program is running on port: ${port}`);
+app.listen(port, async () => {
+  try {
+    console.log(`Program is running on port: ${port}`);
```
Comment (Miszczu2137) [inline src/app.ts]: Rather than console.log log, I'd use some logger, for instance bunyan.

File: src/controllers/fake-controller.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import { Request, Response } from 'express';
+import { IdInfo } from '../contracts/id-info';
+import { DbConnection } from '../db/db-connection';
+import { FakeRepository } from '../db/repositories/fake-repository';
+
+export class FakeController {
+  private readonly fakeRepository: FakeRepository;
+  constructor(connection: DbConnection) {
+    this.fakeRepository = new FakeRepository(connection);
+  }
```
Comment (Miszczu2137) [inline src/controllers/fake-controller.ts:13]: add new line here

File: src/controllers/fake-controller.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import { runInTransaction } from '../db/decorators/run-in-transaction';
+import { Fake } from '../db/entities/fake';
+import { Request, Response } from 'express';
+import { IdInfo } from '../contracts/id-info';
+import { DbConnection } from '../db/db-connection';
+import { FakeRepository } from '../db/repositories/fake-repository';
+
+export class FakeController {
+  private readonly fakeRepository: FakeRepository;
+  constructor(connection: DbConnection) {
```
Comment (Miszczu2137) [inline src/controllers/fake-controller.ts]: add new line here

File: src/db/db-client.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
+import { Pool, PoolClient } from 'pg';
+import config from 'config';
+
+export class DbInstance {
```
Comment (Miszczu2137) [inline src/db/db-client.ts]: To remove coupling between db-client and node-confing you should pass connection string by the DBClient constructor.

File: src/db/db-connection.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    id: number,
+    tableName: string,
+    columns: string[]
+  ): Promise<unknown> {
+    const result = await this.query(
+      `SELECT ${columns.join(',')} FROM public.${tableName} WHERE id=$1;`,
+      [id.toString()]
+    );
+    if (!result[0]) {
+      throw Error('ni mo takiego');
```
Comment (Miszczu2137) [inline src/db/db-connection.ts]: ??

File: Dockerfile (hunk 1 above)
Comment (aziarnik) [reply Dockerfile]: I have splitted it to 3 stages - now there is no duplication

File: config/default.json (hunk 2 above)
Comment (aziarnik) [reply config/default.json]: done

File: config/default.json (hunk 3 above)
Comment (aziarnik) [reply config/default.json]: done

File: docker-compose.debug.yml (hunk 4 above)
Comment (aziarnik) [reply docker-compose.debug.yml]: It was auto-generated by VS Code. I have change debug -> dev

File: package.json (hunk 5 above)
Comment (aziarnik) [reply package.json]: done

File: src/app.ts (hunk 6 above)
Comment (aziarnik) [reply src/app.ts]: done

File: src/controllers/fake-controller.ts (hunk 7 above)
Comment (aziarnik) [reply src/controllers/fake-controller.ts:13]: done

File: src/controllers/fake-controller.ts (hunk 8 above)
Comment (aziarnik) [reply src/controllers/fake-controller.ts]: done

File: src/db/db-client.ts (hunk 9 above)
Comment (aziarnik) [reply src/db/db-client.ts]: done

File: src/db/db-connection.ts (hunk 10 above)
Comment (aziarnik) [reply src/db/db-connection.ts]: Here I was testing async error handling with some dummy data and forgot to change it. It is fixed now

File: docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,29 @@
+version: '3.4'
```
Comment (Miszczu2137) [inline docker-compose.yml]: docker-compose is only for dev purpose, we don't need prod container here.

**Review by Miszczu2137 — APPROVED** [review]

**Pull request merged**